feat : 슬라이더 구현

KimDongGyun23 opened the pull request:
## 변경 사항

- 슬라이더 구현
- TimeSlider, WeightSlider

<br/>

## Store 사용
### timeStore

상태 값
- selectedTime
  - 선택된 시간을 나타내며, 초기 값은 `"00:00:00"` 입니다.
  - 선택되는 시, 분에 따라 해당 값이 업데이트 됩니다.
  - 폼 전송시에 사용되는 형식에 맞도록 설정하였습니다.

<br/>

액션
- handleHourChange, handleMinuteChange
  - 선택되는 숫자를 "00" 형태를 맞추고, 기존 `selectedTime`에 맞추어 업데이트합니다.

<br/>

내보내기
- useSelectedTime
  - 현재 선택된 시간 `selectedTime`을 내보냅니다.
  - ```const time = useSelectedTime()``` 로 사용 가능합니다.

- useSelectedTimeActions
  - 구현된 액션들을 한번에 내보냅니다.
  - ```const { handleHourChange, handleMinuteChange } = useSelectedTimeActions()``` 로 사용 가능합니다.

<br/><br/>

### weightStore
상태 값
- weight
  - 선택된 몸무게를 나타내며, 초기 값은 `0.0` 입니다.
  - 선택되는 십의 자리, 일의 자리, 소수점 자리에 따라 해당 값이 업데이트 됩니다.
  - 폼 전송시에 사용되는 형식에 맞도록 설정하였습니다.

<br/>

액션
- handleTensWeightChange, handleOnesWeightChange, handleDecimalWeightChange
  - 각각 십의 자리, 일의 자리, 소수점 자리를 나타내며 , 기존 `weight`를 업데이트합니다.
  - 소수점 첫째 자리까지 나타내도록 설정하였습니다.

<br/>

내보내기
- useSelectedWeight
  - 현재 선택된 몸무게 `weight`를 내보냅니다.
  - ```const weight = useSelectedWeight()``` 로 사용 가능합니다.

- useSelectedWeightActions
  - 구현된 액션들을 한번에 내보냅니다.
  - ```const { handleTensWeightChange, handleOnesWeightChange, handleDecimalWeightChange } = useSelectedWeightActions()``` 로 사용 가능합니다.

<br/><br/>

## 리뷰 필요

- Swiper 패키지를 활용하여 슬라이더 컴포넌트를 구현했습니다.

- 리렌더링 문제를 해결하기 위해 memo와 디바운스를 사용했습니다. memo는 불필요한 리렌더링을 방지하고, 디바운스는 슬라이드를 넘길 때 발생하는 연속적인 로직 실행을 방지하기 위해 적용하였습니다.

- 디바운스 적용을 위하여 useDebounceFunc 훅을 직접 구현하여 사용했습니다. 콜백 함수와 지연 시간을 전달 받아 디바운스를 적용합니다. 이를 통해 슬라이드를 계속 넘길 때는 로직이 실행되지 않고, 사용자가 슬라이드를 멈췄을 때에만 원하는 로직이 실행되도록 했습니다.

- 몸무게와 시간을 저장하기 위해 zustand의 store를 활용했습니다. 내부 로직을 통해 상태를 서버 API 요청 시 요구되는 형식 ( "HH:mm:ss" | "xx.x" ) 으로 저장하였습니다. 저장된 상태값은 필요한 곳에서 가져올 수 있으며, 추가 로직 없이 올바른 형식으로 서버에 요청을 보낼 수 있을 것입니다.

- 초기 값을 "00:00" | "00.0 kg" 으로 설정하는 것보다, 중앙 값으로 설정하는 것이 유저 입장에서 조절하기 편리할 것이라고 생각했습니다. 이에 따라 중앙 값인 "12:30" | "55.5kg" 을 초기 값으로 설정하였습니다.


close #42

KimDongGyun23 committed:
chore : swiper 패키지 설치

#42

KimDongGyun23 committed:
feat : slider UI 구현

#42

KimDongGyun23 committed:
feat : TimeSlider UI 구현

#42

KimDongGyun23 committed:
feat : useDebounceFunc 함수 구현

#42

KimDongGyun23 committed:
feat : Slider에 useDebounceFunc 적용

#42

KimDongGyun23 committed:
feat : SwiperSlide를 memo를 이용하여 최적화

#42

KimDongGyun23 committed:
chore : SliderType에서 SliderProps로 타입 이름 변경

#42

KimDongGyun23 committed:
chore : Swiper 내부 숫자 너비 설정 및 텍스트 가운데 정렬

#42

KimDongGyun23 committed:
feat : WeightSlider 구현

#42

KimDongGyun23 committed:
feat : WeightStore 생성

몸무게를 관리하는 zustand store 생성

#42

KimDongGyun23 committed:
feat : WeightSlider 컴포넌트에 전역 상태 로직 추가

#42

KimDongGyun23 committed:
feat : timeStore 생성

TimeSlider로 선택된 시간을 전역으로 상태 관리

#42

KimDongGyun23 committed:
feat : TimeSlider에 timeStore 적용

#42

KimDongGyun23 committed:
chore : 변수명 수정 및 클라이언트 컴포넌트 선언

weightSlider 및 Store에서 사용하는 변수명 일부 수정
TimeSlider를 클라이언트 컴포넌트로 선언

#42

KimDongGyun23 committed:
feat : swiper 초기값을 중앙값으로 설정

#42

KimDongGyun23 committed:
chore : 슬라이더 스타일 추가

전체 너비 설정
유저 텍스트 선택 방지

#42

KimDongGyun23 committed:
feat : 슬라이더 스토리북 추가

#42

KimDongGyun23 committed:
Merge branch 'develop' into feature/#42_slider

KimDongGyun23 committed:
fix : TimeSlider ":" 위치 변경

#42

kimsuyeon0916 commented on src/hooks/useDebounceFunc.ts:
useDebounce로 지어도 좋을 것 같아요!

KimDongGyun23 replied:
수연님께서 설명하신 `(value: T)` 를 넘겨주는 디바운스 훅을 `useDebounce`로 많이들 사용한다고 알고있습니다.

혹시 이후에 해당 훅을 구현하게 된다면, 이와 구분하기 위하여 `useDebounceFunc`라 이름 지었습니다.
지금 생각해보니 `useDebounceCallback` 이 조금 더 직관적일 수 있겠네요.

kimsuyeon0916 commented on src/components/slider/TimeSlider.tsx:
```diff
@@ -0,0 +1,22 @@
+import { Slider } from './Slider'
+
+export const TimeSlider = () => {
+  const HOURS = Array.from({ length: 24 }, (_, i) => String(i).padStart(2, '0'))
+  const MINUTES = Array.from({ length: 60 }, (_, i) => String(i).padStart(2, '0'))
```
해당 부분은 변수같아서 대문자보다는 소문자를 사용하는게 좋을 것 같아요!

KimDongGyun23 replied:
다른 값에 영향을 받지않고 절대적으로 변경되지 않는다는 것을 표현하기 위하여 대문자를 사용하였습니다.
혹시 소문자로 표현해야만 하는 이유가 있을까요?

kimsuyeon0916 replied:
i 를 이용해서 배열 처리를 하는 것이 상수에는 적합하지 않다고 판단했었습니다! 배열의 값을 변경할 수도 있고요. Object.freeze() 를 사용해서 배열 값을 변경 못하게 하면 상수라고 판단할 수 있을 것 같아요! 저는 원시값(혹은 변경 불가능한 값)을 상수로 취급하곤 했었던지라 메서드를 사용해서 반환하는걸 상수라고 생각하지 않았던 것 같습니다.

KimDongGyun23 replied:
네 그럼 그 방식으로 고쳐두겠습니다.

kimsuyeon0916 commented on src/hooks/useDebounceFunc.ts:
```diff
@@ -0,0 +1,17 @@
+'use client'
+import { useCallback, useRef } from 'react'
+
+export type UseDebounceFuncType = (callback: () => void, delay: number) => () => void
```
어떤 값을 디바운스 시키기 위한 공통 훅은 callback 함수를 매개변수로 받기보단 value: T 를 받고 콜백함수는 내부에서 정의해서 활용하는 것이 좋을 것 같은데, 동균님의 의견이 궁금합니다!

KimDongGyun23 replied:
질문에 대해서 어떻게 이야기를 해야할 지 고민되었습니다. 
물론 공통 커스텀 훅이 여러 곳에서 사용할 수는 있어야 하지만, 특정 형식에 맞춰야 할 이유는 없다고 생각합니다.

<br/>

저 또한 value를 넘겨 받아 사용하는 디바운스 함수를 사용하기도 합니다.
하지만 해당 로직에서는 callback 함수를 넘겨받는 것이 훨씬 효율적이라고 생각하였고, 다른 곳에서도 충분히 활용 가능하다고 판단하여 구현하였습니다.

<br/>

말씀 하신 형식을 Slider 로직에 적용한다면, 인덱스 값의 변경을 디바운스로 처리할 것 같네요.
디바운스된 값이 변경될 때마다 onChange를 실행할텐데, 그럴려면 useEffect를 사용해야 하지 않을까요? 
useEffect를 사용한다면 변경 시점마다 렌더링이 될 것이라고 생각합니다.

<br/>

Slider에서는 변경된 값을 저장하기만 하므로 작성한 디바운스 로직으로 리렌더링을 방지할 수 있다고 생각하였습니다.

<br/>

개인적으로 커스텀 훅을 사용할 때에는 왜 사용하는지, 어떤 부분에서 효율성을 얻을 수 있는지를 많이 고민하는 것 같습니다. 해당 훅 또한 여러 방면에서 고민을 많이 해보았는데 더 좋은 방안이 있다면 수정해보겠습니다.

kimsuyeon0916 replied:
넵 알겠습니다! 참고로 onSlideChange props 로 들어가는 함수에서 realIndex 는 굳이 저희가 관리 안해도 되지 않나요? debounce 안해도 onChange 를 쌩으로 넣어도 리렌더링이 일어나지 않길래 여쭤봅니다!

KimDongGyun23 replied:
Swiper를 사용하였을 때 realIndex가 있어야 정확한 현재의 값을 설정할 수 있습니다.

kimsuyeon0916 replied:
넵넵~! 근데 debounce 를 안하고 onChange 를 넣어도 리렌더링이 일어나지 않길래, realIndex 자체는 라이브러리 내부에서 관리하니까 사용자가 슬라이드를 멈췄을 때에만 원하는 로직이 실행되도록 하신다는게 어떤 로직인지가 궁금했습니다!

KimDongGyun23 replied:
넵 무슨 말씀인지 이해했습니다. 내일 오후 전까지 확인해보고 코멘트 남기겠습니다.

KimDongGyun23 replied:
아직 이동 중이라 확인은 못했지만, 제가 생각했던 로직에 대해서 적어봅니다.
예를 들어 onChange가 콘솔에 현재 슬라이드 값을 출력한다는 함수라고 해보죠.

제가 원했던 로직은 슬라이드가 멈췄을 때에만 콘솔이 찍히는 것이었습니다.
그렇지 않다면 넘어가는 모든 슬라이드의 값들이 콘솔에 출력될 것입니다.

실제로 디바운스 적용 전에는 콘솔에 지나간 모든 값들이 찍히는 것을 확인했었고, 이를 개선하기 위해 디바운스를 도입했었습니다.

리렌더링보다는 함수의 실행 시점에 관해 생각하고 도입했었습니다.
저도 다시 한번 확인해보겠습니다. :)

kimsuyeon0916 commented on src/hooks/useDebounceFunc.ts:
```diff
@@ -0,0 +1,17 @@
+export type UseDebounceFuncType = (callback: () => void, delay: number) => () => void
+
+export const useDebounceFunc: UseDebounceFuncType = (callback, delay) => {
+  const timerRef = useRef<NodeJS.Timeout | undefined>(undefined)
```
useEffect 대신 useRef 를 활용해서 구현하신 이유가 궁금합니다!

KimDongGyun23 replied:
딱히 별다른 이유는 없습니다. 개인 스타일에 따른 차이라고 생각하는데 혹시 다른 SideEffect가 존재하나요?

kimsuyeon0916 replied:
보편적으로는 useEffect를 사용한 코드를 많이 접했던지라 useRef 로 구현하신 이유가 궁금했습니다!

kimsuyeon0916 commented:
파일 몇몇에서 interface 로 정의된 부분들이 있습니다! 저희 컨벤션으로 type 사용하기로 했었으니까 해당 부분 수정해주시면 감사할 것 같습니당~!

kimsuyeon0916 commented on src/components/slider/Slider.tsx:
```diff
@@ -0,0 +1,54 @@
+  content: string
+}
+
+const MemoizedSwiperSlide = memo(({ content }: MemoizedSwiperSlideProps) => {
+  return <SwiperSlide className="leading-[30px]">{content}</SwiperSlide>
+})
```
동균님은 메모이제이션을 이용하실 때 어떤 기준으로 이용해야겠다고 생각하시나요!? 리렌더링 방지를 위해서 memo가 꼭 필요한 이유가 궁금합니다.

KimDongGyun23 replied:
Slider를 구현할 때, 우선 swiper를 활용하여 UI를 구현하였습니다. 

스와이프를 할 때마다 리렌더링이 많이 발생하였고, 어떤 부분에서 렌더링이 발생하는 것인지 검사해보았습니다. 
이때, SwiperSlide에 해당하는 부분(각각의 숫자)들이 문제라는 것을 확인했습니다.

각각의 `<SwiperSlide>`는 동적으로 변경되는 값이 없으며 단순한 UI만 보여준다고 판단하였습니다.
그래서 해당 부분을 처음에만 불러오면 될 것이라고 생각하였고, 이를 memo로 감쌌더니 리렌더링 문제가 해결되는 것을 확인할 수 있었습니다.

저도 최적화 부분은 공부가 많이 필요합니다. 
최대한 UI적으로 리렌더링 되는 부분들은 memo를 활용하여 해결하고자 합니다.

혹시 다른 의견이나 더 좋은 방법이 있을까요?

kimsuyeon0916 replied:
저는 우선적으로는 컴포넌트 분할로 리렌더링을 해결하곤 합니다..! 근데 이미 MemoizedSwiperSlide 로 분리를 해놓으셔서 memo를 제거해도 리렌더링이 일어나지 않는 것 같은데 memo를 사용하신 이유가 궁금했습니다! 한번 확인해보시면 좋을 것 같습니다!

KimDongGyun23 replied:
네 알겠습니다.

KimDongGyun23 committed:
chore : interface 에서 type으로 수정

#42

KimDongGyun23 committed:
chore : useDebounceFunc 이름 변경 및 hooks/index.ts로 내보내기

useDebounceFunc에서 useDebounceCallback로 이름 변경 및 배럴 파일 내보내기

#42

KimDongGyun23 committed:
chore : slide 상수 수정

상수 파일로 이동 후, Object.freeze로 감싸서 내보내기
Slider 컴포넌트에서 readonly string[]으로 내려받기

#42

G0MTENG commented on src/components/slider/TimeSlider.stories.tsx:
<img width="124" alt="스크린샷 2024-08-28 오전 8 19 49" src="https://github.com/user-attachments/assets/6ceb7a8e-f2e2-4f42-a4bc-9322ec9a2082">

TimeSlider와 WeightSlider 이렇게 두 개의 컴포넌트를 보내는 것보다 하나로 감싸서

<img width="125" alt="스크린샷 2024-08-28 오전 8 20 40" src="https://github.com/user-attachments/assets/7cca056c-1052-416b-82f4-c96710675a09">

이러한 방식을 사용하는건 어떠신가요 ?

```js
import type { PropsWithChildren } from 'react'
import type { Meta, StoryFn } from '@storybook/react'

import { TimeSlider } from './TimeSlider'
import { WeightSlider } from './WeightSlider'

const SliderWrapper = ({ children }: PropsWithChildren) => {
  return <div className="w-[450px]">{children}</div>
}

export default {
  title: 'Slider',
  component: SliderWrapper,
} as Meta

export const Time: StoryFn<typeof TimeSlider> = () => {
  return (
    <SliderWrapper>
      <TimeSlider />
    </SliderWrapper>
  )
}

export const Weight: StoryFn<typeof WeightSlider> = () => {
  return (
    <SliderWrapper>
      <WeightSlider />
    </SliderWrapper>
  )
}
```

다음과 같이 두 컴포넌트를 children로 받는 랩핑 컴포넌트를 선언하고, 등록해서 사용하면 Slider라는 공통 주제로 묶어볼 수 있을거 같습니다!!

KimDongGyun23 replied:
네 좋습니다. 좋은 방법인 것 같아요!

KimDongGyun23 replied:
수정해서 올렸습니다.

KimDongGyun23 committed:
chore : TimeSlider와 WeightSlider의 스토리북을 합치기

#42
